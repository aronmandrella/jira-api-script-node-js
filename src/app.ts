import { Command, CommanderError } from "@commander-js/extra-typings";
import chalk from "chalk";
import { JiraAPI } from "./JiraAPI";

export async function getJiraComponentsWithoutLead(query: {
	baseUrl: string;
	projectId: string;
}): Promise<{ id: string; name: string; issues: number }[]> {
	const { baseUrl, projectId } = query;

	const api = new JiraAPI({ baseUrl, projectId });

	const components = await api.getComponents();

	const componentsWithoutLead = components.filter((component) => !component.lead);
	if (componentsWithoutLead.length === 0) {
		return [];
	}

	const issues = await api.getIssuesByComponents({
		componentIds: componentsWithoutLead.map((component) => component.id)
	});

	const issuesCountByComponentId = new Map<string, number>();

	for (const issue of issues) {
		for (const component of issue.fields.components) {
			issuesCountByComponentId.set(
				component.id,
				(issuesCountByComponentId.get(component.id) ?? 0) + 1
			);
		}
	}

	return componentsWithoutLead.map((component) => {
		return {
			id: component.id,
			name: component.name,
			issues: issuesCountByComponentId.get(component.id) ?? 0
		};
	});
}

/**
 * Entry point of the CLI ap. It can be used like this:
 * ```sh
 * node app.js --help
 * node app.js --jira-base-url https://abc.atlassian.net --jira-project-id xyz
 * ```
 */
export async function app() {
	try {
		const program = new Command()
			.name("app.js")
			.requiredOption("--jira-base-url <string>")
			.requiredOption("--jira-project-id <string>")
			.showHelpAfterError()
			.exitOverride()
			.configureOutput({
				writeOut: (str) => {
					console.log(str.trim());
				},
				writeErr: (str) => {
					console.error(str.trim());
				},
				outputError: (str, write) => {
					write(chalk.red(str));
				}
			});

		console.log(chalk.cyan.bold("Detecting Jira components without a component lead..."));
		console.log();

		const { jiraBaseUrl, jiraProjectId } = program.parse().opts();

		console.log("• Jira base url:", chalk.bold(jiraBaseUrl));
		console.log("• Jira project: ", chalk.bold(jiraProjectId));
		console.log();

		const components = await getJiraComponentsWithoutLead({
			baseUrl: jiraBaseUrl,
			projectId: jiraProjectId
		});

		if (components.length === 0) {
			console.log(
				chalk.green("Script didn't detect any components without a project lead :)")
			);
		} else {
			console.log(
				chalk.yellow(
					`Script detected ${chalk.bold(
						components.length
					)} component(s) without a project lead:`
				)
			);
			console.log();

			for (const { id, name, issues } of components) {
				console.log(
					chalk.gray(`[ID: ${id}]`),
					chalk.bold(name),
					chalk.magenta(`with ${chalk.bold(issues)} issue(s)`)
				);
			}
		}

		process.exitCode = 0;
	} catch (error) {
		if (error instanceof CommanderError) {
			console.error();
		} else {
			console.error(error);
			console.error();
		}

		console.error(chalk.red("Oops, something went wrong."));
		console.error(chalk.red("The detailed error can be found above this message."));

		process.exitCode = 1;
	}
}

// eslint-disable-next-line unicorn/prefer-module
if (require.main === module) {
	// eslint-disable-next-line @typescript-eslint/no-floating-promises
	app();
}
