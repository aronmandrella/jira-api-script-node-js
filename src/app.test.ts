import os from "node:os";
import {
	afterAll,
	afterEach,
	beforeEach,
	describe,
	expect,
	it,
	jest
} from "@jest/globals";
import fetchMock from "fetch-mock";
import stripAnsi from "strip-ansi";
import ExampleGetComponentsResponse from "./__fixtures__/GetComponentsResponse.json";
import ExampleGetIssuesByComponentsResponse from "./__fixtures__/GetIssuesByComponentsResponse.json";
import { app } from "./app";

const VALID_CLI_ARGUMENTS: Record<string, string> = {
	"--jira-base-url": "https://herocoders.atlassian.net",
	"--jira-project-id": "SP"
};

const INVALID_CLI_ARGUMENTS: Record<string, string> = {
	"--jira-base-url": "https://xxx.atlassian.net",
	"--jira-project-id": "XXX"
};

function setArgv(args: string[]) {
	process.argv = ["node.exe", "app.js", ...args];
}

describe("app", () => {
	let stderr: string;
	let stdout: string;

	jest.spyOn(console, "log").mockImplementation((...data) => {
		stdout += stripAnsi(data.map(String).join(" ") + os.EOL);
	});

	jest.spyOn(console, "error").mockImplementation((...data) => {
		stderr += stripAnsi(data.map(String).join(" ") + os.EOL);
	});

	beforeEach(() => {
		stderr = "";
		stdout = "";
		setArgv([]);
	});

	afterEach(() => {
		fetchMock.restore();
	});

	afterAll(() => {
		jest.restoreAllMocks();
	});

	describe("e2e", () => {
		it.each(Object.keys(VALID_CLI_ARGUMENTS))(
			"should fail when '%s' CLI argument is missing",
			async (argumentName) => {
				expect.hasAssertions();

				const filteredArgs = Object.entries(VALID_CLI_ARGUMENTS)
					.filter(([key]) => {
						return key !== argumentName;
					})
					.flat();

				setArgv(filteredArgs);
				await expect(app()).resolves.toBeUndefined();
				expect(process.exitCode).toBe(1);
				expect(stderr).toContain(`error: required option '${argumentName}`);
			}
		);

		it.each(Object.keys(VALID_CLI_ARGUMENTS))(
			"should fail when '%s' CLI argument is invalid",
			async (argumentName) => {
				expect.hasAssertions();

				const corruptedArgs = Object.entries({
					...VALID_CLI_ARGUMENTS,
					[argumentName]: String(INVALID_CLI_ARGUMENTS[argumentName])
				}).flat();

				setArgv(corruptedArgs);
				await expect(app()).resolves.toBeUndefined();
				expect(process.exitCode).toBe(1);
				expect(stderr).toContain(`JiraResponseError`);
			}
		);

		it("should succeed when CLI arguments are valid", async () => {
			expect.hasAssertions();

			setArgv(Object.entries(VALID_CLI_ARGUMENTS).flat());
			await expect(app()).resolves.toBeUndefined();
			expect(process.exitCode).toBe(0);

			expect(stdout).toMatchSnapshot();
		});
	});

	describe("mocked", () => {
		const MOCK_CLI_ARGUMENTS = INVALID_CLI_ARGUMENTS;

		const baseUrl = String(MOCK_CLI_ARGUMENTS["--jira-base-url"]);
		const projectId = String(MOCK_CLI_ARGUMENTS["--jira-project-id"]);

		it("should fail when API responses are not valid", async () => {
			expect.hasAssertions();

			fetchMock.mock(`${baseUrl}/rest/api/3/project/${projectId}/components`, {
				status: 400
			});
			fetchMock.mock(`begin:${baseUrl}/rest/api/3/search`, {
				status: 400
			});

			setArgv(Object.entries(MOCK_CLI_ARGUMENTS).flat());
			await expect(app()).resolves.toBeUndefined();
			expect(process.exitCode).toBe(1);
		});

		it("should succeed when API responses are valid", async () => {
			expect.hasAssertions();

			fetchMock.mock(`${baseUrl}/rest/api/3/project/${projectId}/components`, {
				status: 200,
				body: ExampleGetComponentsResponse
			});

			fetchMock.mock(`begin:${baseUrl}/rest/api/3/search`, {
				status: 200,
				body: ExampleGetIssuesByComponentsResponse
			});

			setArgv(Object.entries(MOCK_CLI_ARGUMENTS).flat());
			await expect(app()).resolves.toBeUndefined();
			expect(process.exitCode).toBe(0);

			expect(stdout).toMatchSnapshot();
		});

		it("should correctly handle a case without components", async () => {
			expect.hasAssertions();

			fetchMock.mock(`${baseUrl}/rest/api/3/project/${projectId}/components`, {
				status: 200,
				body: []
			});

			fetchMock.mock(`begin:${baseUrl}/rest/api/3/search`, {
				status: 200,
				body: ExampleGetIssuesByComponentsResponse
			});

			setArgv(Object.entries(MOCK_CLI_ARGUMENTS).flat());
			await expect(app()).resolves.toBeUndefined();
			expect(process.exitCode).toBe(0);

			expect(stdout).toMatchSnapshot();
		});
	});
});
