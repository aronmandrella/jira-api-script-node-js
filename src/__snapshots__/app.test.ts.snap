// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app e2e should succeed when CLI arguments are valid 1`] = `
"Detecting Jira components without a component lead...

• Jira base url: https://herocoders.atlassian.net
• Jira project:  SP

Script detected 3 component(s) without a project lead:

[ID: 10130] Backend with 1 issue(s)
[ID: 10128] Synchronization with 2 issue(s)
[ID: 10131] Templates with 5 issue(s)
"
`;

exports[`app mocked should correctly handle a case without components 1`] = `
"Detecting Jira components without a component lead...

• Jira base url: https://xxx.atlassian.net
• Jira project:  XXX

Script didn't detect any components without a project lead :)
"
`;

exports[`app mocked should succeed when API responses are valid 1`] = `
"Detecting Jira components without a component lead...

• Jira base url: https://xxx.atlassian.net
• Jira project:  XXX

Script detected 3 component(s) without a project lead:

[ID: 10130] Backend with 1 issue(s)
[ID: 10128] Synchronization with 2 issue(s)
[ID: 10131] Templates with 5 issue(s)
"
`;
